File: CharAnalysis_2_0_MATLAB/CharValidateParams.m
function CharValidateParams(charData, Pretreatment, Smoothing, ...
    PeakAnalysis, Results)
% CharValidateParams   Validate CharAnalysis input parameters before run.
%   CharValidateParams(charData, Pretreatment, Smoothing, PeakAnalysis,
%       Results)
%
%   Checks all user-supplied parameters for internal consistency and
%   plausible ranges. Throws a descriptive error and halts execution if
%   any check fails. Emits warnings for non-fatal but unusual settings.
%
%   Called from CharAnalysis.m immediately after CharParameters returns,
%   replacing the two inline if/error blocks that existed in v1.1.
%
%   Checks performed
%   ────────────────
%   charData
%     1. At least 6 columns present (cmTop cmBot ageTop ageBot charVol charCount)
%     2. No NaN values in the age columns (cols 3 and 4)
%     3. Ages are monotonically non-decreasing (ageTop col 3)
%
%   Pretreatment
%     4. zoneDiv has at least 2 values
%     5. zoneDiv is strictly ascending
%     6. yrInterp is non-negative (0 = auto)
%     7. zoneDiv(end) does not exceed the bottom age of the last raw sample
%        (warning only -- v2.0 marks overrun samples NaN, but results will
%        differ from v1.1 which silently filled them with zero CHAR)
%
%   Smoothing
%     8. Smoothing window (yr) is shorter than the record length
%     9. Smoothing method is in the range 1-5
%
%   PeakAnalysis
%     10. threshType is 1 or 2
%     11. threshMethod is 1, 2, or 3
%     12. Locally defined threshold cannot be user-defined
%         (threshMethod == 1 && threshType == 2)
%     13. threshValues all in (0, 1) when threshMethod > 1
%     14. cPeak is 1 (residuals) or 2 (ratios)
%     15. minCountP is in [0, 1]
%     16. peakFrequ is positive
%
%   CharAnalysis v2.0

%% ── 1. charData column count ─────────────────────────────────────────────
if size(charData, 2) < 6
    error('CharValidateParams: charData must have at least 6 columns (cmTop, cmBot, ageTop, ageBot, charVol, charCount). Found %d.', ...
        size(charData, 2))
end

%% ── 2. No NaN ages ───────────────────────────────────────────────────────
if any(isnan(charData(:,3))) || any(isnan(charData(:,4)))
    error('CharValidateParams: charData contains NaN values in the age columns (cols 3–4). All sample ages must be specified.')
end

%% ── 3. Ages monotonically non-decreasing ─────────────────────────────────
ageTops = charData(:,3);
if any(diff(ageTops) < 0)
    badRows = find(diff(ageTops) < 0);
    error('CharValidateParams: ageTop values (col 3) are not monotonically non-decreasing. First violation at row %d.', ...
        badRows(1))
end

%% ── 4. zoneDiv has at least 2 values ─────────────────────────────────────
if numel(Pretreatment.zoneDiv) < 2
    error('CharValidateParams: zoneDiv must contain at least 2 values (start and end of record).')
end

%% ── 5. zoneDiv strictly ascending ───────────────────────────────────────
if any(diff(Pretreatment.zoneDiv) <= 0)
    error('CharValidateParams: zoneDiv must be strictly ascending (youngest age first).')
end

%% ── 6. yrInterp non-negative ─────────────────────────────────────────────
if Pretreatment.yrInterp < 0
    error('CharValidateParams: yrInterp must be >= 0 (0 = use median sample resolution).')
end

%% ── 7. zoneDiv(end) within data extent (warning) ────────────────────────
% If zoneDiv(end) exceeds the bottom age of the last raw sample, the
% vectorized proportion matrix in CharPretreatment will find no overlapping
% raw data for those terminal intervals and mark them NaN. This causes
% charBkg to differ from v1.1, which silently filled them with zero CHAR.
% Set zoneDiv(end) to no greater than the bottom age of the last raw sample.
lastAgeBotInData = charData(end, 4);
if Pretreatment.zoneDiv(end) > lastAgeBotInData
    warning(['CharValidateParams: zoneDiv(end) (%g yr BP) exceeds the ' ...
             'bottom age of the last raw sample (%g yr BP). ' ...
             'Terminal interpolated samples beyond the data will be ' ...
             'set to NaN. Set zoneDiv(end) <= %g to avoid this.'], ...
        Pretreatment.zoneDiv(end), lastAgeBotInData, lastAgeBotInData)
end

%% ── 8. Smoothing window shorter than record ──────────────────────────────
recordLength = max(Pretreatment.zoneDiv) - min(Pretreatment.zoneDiv);
smoothYr     = Smoothing.yr(end);   % Use the longest window specified.
if smoothYr >= recordLength
    error('CharValidateParams: smoothing window (%d yr) must be shorter than the record length (%d yr).', ...
        smoothYr, recordLength)
end

%% ── 9. Smoothing method in range ─────────────────────────────────────────
if ~ismember(Smoothing.method, 1:5)
    error('CharValidateParams: smoothing method must be 1–5. Got %d.', ...
        Smoothing.method)
end

%% ── 10. threshType valid ──────────────────────────────────────────────────
if ~ismember(PeakAnalysis.threshType, [1 2])
    error('CharValidateParams: threshType must be 1 (global) or 2 (local). Got %d.', ...
        PeakAnalysis.threshType)
end

%% ── 11. threshMethod valid ───────────────────────────────────────────────
if ~ismember(PeakAnalysis.threshMethod, [1 2 3])
    error('CharValidateParams: threshMethod must be 1, 2, or 3. Got %d.', ...
        PeakAnalysis.threshMethod)
end

%% ── 12. Local threshold cannot be user-defined ───────────────────────────
if PeakAnalysis.threshMethod == 1 && PeakAnalysis.threshType == 2
    error('CharValidateParams: a locally defined threshold (threshType = 2) cannot be user-defined (threshMethod = 1). Change input parameters.')
end

%% ── 13. threshValues in (0,1) when threshMethod > 1 ─────────────────────
if PeakAnalysis.threshMethod > 1
    tv = PeakAnalysis.threshValues(1:4);
    if any(tv <= 0) || any(tv >= 1)
        error('CharValidateParams: all threshValues must be in the open interval (0, 1) when threshMethod = %d. Got [%s].', ...
            PeakAnalysis.threshMethod, num2str(tv))
    end
end

%% ── 14. cPeak valid ──────────────────────────────────────────────────────
if ~ismember(PeakAnalysis.cPeak, [1 2])
    error('CharValidateParams: cPeak must be 1 (residuals) or 2 (ratios). Got %d.', ...
        PeakAnalysis.cPeak)
end

%% ── 15. minCountP in [0,1] ───────────────────────────────────────────────
if PeakAnalysis.minCountP < 0 || PeakAnalysis.minCountP > 1
    error('CharValidateParams: minCountP must be in [0, 1]. Got %g.', ...
        PeakAnalysis.minCountP)
end

%% ── 16. peakFrequ positive ───────────────────────────────────────────────
if PeakAnalysis.peakFrequ <= 0
    error('CharValidateParams: peakFrequ must be positive. Got %g.', ...
        PeakAnalysis.peakFrequ)
end

%% ── Non-fatal warnings ───────────────────────────────────────────────────
% Warn if smoothing window is very short relative to record length.
if smoothYr < 0.05 * recordLength
    warning('CharValidateParams: smoothing window (%d yr) is less than 5%% of the record length (%d yr). Consider a longer window.', ...
        smoothYr, recordLength)
end

% Warn if local threshold window may contain fewer than 30 samples.
if PeakAnalysis.threshType == 2 && Pretreatment.yrInterp > 0
    samplesInWindow = smoothYr / Pretreatment.yrInterp;
    if samplesInWindow < 30
        warning('CharValidateParams: smoothing window (%d yr) at yrInterp = %d yr gives ~%d samples per local threshold window. At least 30 are recommended.', ...
            smoothYr, Pretreatment.yrInterp, floor(samplesInWindow))
    end
end

disp('      Parameter validation passed.')
end

File: CharAnalysis_2_0_R/R/CharValidateParams.R
#' Validate CharAnalysis input parameters
#'
#' Checks all user-supplied parameters for internal consistency and plausible
#' ranges.  Throws a descriptive error and halts if any check fails.  Emits
#' warnings for non-fatal but unusual settings.
#'
#' Mirrors \code{CharValidateParams.m} from the MATLAB v2.0 codebase.  All 15
#' checks and 2 non-fatal warnings are reproduced in the same order.
#'
#' @param char_data    Numeric matrix (n x 6+).
#' @param pretreatment List with \code{zoneDiv}, \code{yrInterp},
#'   \code{transform}.
#' @param smoothing    List with \code{method}, \code{yr}.
#' @param peak_analysis List with \code{cPeak}, \code{threshType},
#'   \code{threshMethod}, \code{threshValues}, \code{minCountP},
#'   \code{peakFrequ}.
#' @param results      List (not checked; included for API symmetry).
#'
#' @return \code{NULL} invisibly on success.  Stops with an error on failure.
#'
#' @seealso [char_parameters()], [CharAnalysis()]
char_validate_params <- function(char_data, pretreatment, smoothing,
                                  peak_analysis, results) {

  # 1. At least 6 columns ---------------------------------------------------
  if (ncol(char_data) < 6L) {
    stop("char_validate_params: char_data must have at least 6 columns ",
         "(cmTop, cmBot, ageTop, ageBot, charVol, charCount). Found ",
         ncol(char_data), ".")
  }

  # 2. No NA ages ------------------------------------------------------------
  if (anyNA(char_data[, 3L]) || anyNA(char_data[, 4L])) {
    stop("char_validate_params: char_data contains NA values in the age ",
         "columns (cols 3-4). All sample ages must be specified.")
  }

  # 3. Ages monotonically non-decreasing ------------------------------------
  age_tops <- char_data[, 3L]
  if (any(diff(age_tops) < 0)) {
    bad_row <- which(diff(age_tops) < 0)[1L]
    stop("char_validate_params: ageTop values (col 3) are not monotonically ",
         "non-decreasing. First violation at row ", bad_row, ".")
  }

  # 4. zoneDiv at least 2 values --------------------------------------------
  if (length(pretreatment$zoneDiv) < 2L) {
    stop("char_validate_params: zoneDiv must contain at least 2 values ",
         "(start and end of record).")
  }

  # 5. zoneDiv strictly ascending -------------------------------------------
  if (any(diff(pretreatment$zoneDiv) <= 0)) {
    stop("char_validate_params: zoneDiv must be strictly ascending ",
         "(youngest age first).")
  }

  # 6. yrInterp non-negative ------------------------------------------------
  if (pretreatment$yrInterp < 0) {
    stop("char_validate_params: yrInterp must be >= 0 ",
         "(0 = use median sample resolution).")
  }

  # 7. Smoothing window shorter than record ---------------------------------
  record_len <- max(pretreatment$zoneDiv) - min(pretreatment$zoneDiv)
  smooth_yr  <- smoothing$yr[length(smoothing$yr)]   # longest window specified
  if (smooth_yr >= record_len) {
    stop("char_validate_params: smoothing window (", smooth_yr,
         " yr) must be shorter than the record length (", record_len, " yr).")
  }

  # 8. Smoothing method in range 1-5 ----------------------------------------
  if (!smoothing$method %in% 1:5) {
    stop("char_validate_params: smoothing method must be 1-5. Got ",
         smoothing$method, ".")
  }

  # 9. threshType valid (1 or 2) --------------------------------------------
  if (!peak_analysis$threshType %in% c(1, 2)) {
    stop("char_validate_params: threshType must be 1 (global) or 2 (local). ",
         "Got ", peak_analysis$threshType, ".")
  }

  # 10. threshMethod valid (1, 2, or 3) -------------------------------------
  if (!peak_analysis$threshMethod %in% c(1, 2, 3)) {
    stop("char_validate_params: threshMethod must be 1, 2, or 3. Got ",
         peak_analysis$threshMethod, ".")
  }

  # 11. Local threshold cannot be user-defined ------------------------------
  if (peak_analysis$threshMethod == 1L && peak_analysis$threshType == 2L) {
    stop("char_validate_params: a locally defined threshold (threshType = 2) ",
         "cannot be user-defined (threshMethod = 1). Change input parameters.")
  }

  # 12. threshValues in (0, 1) when threshMethod > 1 ------------------------
  if (peak_analysis$threshMethod > 1L) {
    tv <- peak_analysis$threshValues[1:4]
    if (any(tv <= 0) || any(tv >= 1)) {
      stop("char_validate_params: all threshValues must be in the open ",
           "interval (0, 1) when threshMethod = ", peak_analysis$threshMethod,
           ". Got [", paste(tv, collapse = ", "), "].")
    }
  }

  # 13. cPeak valid (1 or 2) ------------------------------------------------
  if (!peak_analysis$cPeak %in% c(1, 2)) {
    stop("char_validate_params: cPeak must be 1 (residuals) or 2 (ratios). ",
         "Got ", peak_analysis$cPeak, ".")
  }

  # 14. minCountP in [0, 1] -------------------------------------------------
  if (peak_analysis$minCountP < 0 || peak_analysis$minCountP > 1) {
    stop("char_validate_params: minCountP must be in [0, 1]. Got ",
         peak_analysis$minCountP, ".")
  }

  # 15. peakFrequ positive --------------------------------------------------
  if (peak_analysis$peakFrequ <= 0) {
    stop("char_validate_params: peakFrequ must be positive. Got ",
         peak_analysis$peakFrequ, ".")
  }

  # -- Non-fatal warnings ---------------------------------------------------
  if (smooth_yr < 0.05 * record_len) {
    warning("char_validate_params: smoothing window (", smooth_yr,
            " yr) is less than 5% of the record length (", record_len,
            " yr). Consider a longer window.")
  }

  if (peak_analysis$threshType == 2L && pretreatment$yrInterp > 0) {
    samples_in_window <- smooth_yr / pretreatment$yrInterp
    if (samples_in_window < 30) {
      warning("char_validate_params: smoothing window (", smooth_yr,
              " yr) at yrInterp = ", pretreatment$yrInterp,
              " yr gives ~", floor(samples_in_window),
              " samples per local threshold window. ",
              "At least 30 are recommended.")
    }
  }

  message("      Parameter validation passed.")
  invisible(NULL)
}
